Sources/MongoSwift/ConnectionPool.swift

```diff
@@ -230,12 +250,22 @@ internal class ConnectionPool {
     /// may only be called before any connections are checked out of the pool.** Ideally this code would just live in
     /// `ConnectionPool.init`. However, the client we accept here has to be fully initialized before we can pass it
     /// as the context. In order for it to be fully initialized its pool must exist already.
+    ///
+    /// This method takes ownership of the provided `mongoc_apm_callbacks_t`, so the pointer MUST NOT be used or freed
+    /// after it is passed to this method.
     internal func setAPMCallbacks(callbacks: OpaquePointer, client: MongoClient) {
         // lock isn't needed as this is called before pool is in use.
         switch self.state {
-        case let .open(pool):
-            mongoc_client_pool_set_apm_callbacks(pool, callbacks, Unmanaged.passUnretained(client).toOpaque())
-        case .closing, .closed:
+        case let .opening(future):
+            future.whenSuccess { pool in
```

Comment:
I'm a little worried about this. could we end up with an interleaving like
1) `future` finishes
2) `future.wait()` is unblocked, so a connection can be checked out above, and is immediately
3) the event loop gets around to executing this callback, but the checked-out connection doesn't properly get the APM callbacks set on it

?

Reply:
Great catch, this definitely seems like it could be an issue. I updated `setAPMCallbacks` to use `map` instead of `whenSuccess` to return a new future. Now the calls to `wait` will be on the mapped future and therefore should complete after the callbacks are actually set on the pool.